### ISSUE TITLE
Resume from long pause, Tidal issue

### ISSUE THREAD

**Issue by iamsi** [MusicPlayerDaemon/MPD#241]
if you put a track on pause for a long time (2 hours+ I'm trying to narrow down the shortest time) when you resume it'll play a few seconds then skip to the next track, this is in the log from the resume point:

Feb 28 10:51 : alsa_output: opened hw:0,0 type=HW
Feb 28 10:51 : alsa_output: buffer: size=128..131072 time=2902..2972155
Feb 28 10:51 : alsa_output: period: size=64..65536 time=1451..1486078
Feb 28 10:51 : alsa_output: default period_time = buffer_time/4 = 500000/4 = 125000
Feb 28 10:51 : alsa_output: format=S16_LE (Signed 16 bit Little Endian)
Feb 28 10:51 : alsa_output: buffer_size=22050 period_size=4410
Feb 28 10:51 : output: opened "ALSA default" (alsa) audio_format=44100:16:2
Feb 28 10:51 : state_file: Saving state file /var/lib/mpd/state
Feb 28 10:51 : exception: CURL failed: Recv failure: Connection reset by peer
Feb 28 10:51 : tidal: Tidal track '54144048' resolves to http://57.audio-pop.tidal.com/84044145/d12a9dec856a239abf192c596cf59d8b_26.flac?<snip>
Feb 28 10:51 : decoder_thread: probing plugin flac
Feb 28 10:51 : decoder: audio_format=44100:16:2, seekable=true
Feb 28 10:52 : player: played "tidal://track/18590235"       <--- track that was paused.
Feb 28 10:52 : playlist: queue song 37:"tidal://track/51005870"    
Feb 28 10:52 : state_file: Saving state file /var/lib/mpd/state

I guess the resume from pause logic would need to reload the track and seek to the correct place.

I'm not sure if this is a bug, or a feature request, or just me being crazy.

**Comment by MaxKellermann** [comment]
Those Tidal URLs expire eventually, and 2 hours may be past that point. If you pause for too long, the Tidal web server disconnects MPD, and when MPD attempts to reconnect, it will see the URL has already expired. So yes, we need MPD to re-query the song URL before reconnecting.

**Comment by iamsi** [comment]
it seems "long pause" can be as little as 10 minutes, but constantly it happens after 1 hour.

**Comment by nicolabotta** [comment]
The problem constantly happens on long tracks (10-20 minutes) even without pausing and no matter whether from Tidal or from Qobuz:

Jun 14 03:38 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614334"
Jun 14 03:57 : curl: curl failed: transfer closed with 8833846 bytes remaining to read
Jun 14 03:57 : flac: FLAC__STREAM_DECODER_ABORTED
Jun 14 03:57 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614302"
Jun 14 04:11 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614311"
Jun 14 04:22 : curl: curl failed: transfer closed with 69467230 bytes remaining to read
Jun 14 04:22 : flac: FLAC__STREAM_DECODER_ABORTED
Jun 14 04:22 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614320"
Jun 14 04:44 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614303"
Jun 14 05:01 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614312"
Jun 14 05:08 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614321"
Jun 14 05:19 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614329"
Jun 14 05:32 : curl: curl failed: transfer closed with 24021371 bytes remaining to read
Jun 14 05:32 : flac: FLAC__STREAM_DECODER_ABORTED
Jun 14 05:32 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614304"
Jun 14 05:45 : curl: curl failed: transfer closed with 19180209 bytes remaining to read
Jun 14 05:45 : flac: FLAC__STREAM_DECODER_ABORTED
Jun 14 05:45 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614313"
Jun 14 05:51 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614322"
Jun 14 06:09 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614330"
Jun 14 06:26 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614335"
Jun 14 06:41 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614305"
Jun 14 07:07 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614314"
Jun 14 07:31 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614323"
Jun 14 07:49 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614306"
Jun 14 08:03 : curl: curl failed: transfer closed with 2097865 bytes remaining to read
Jun 14 08:03 : flac: FLAC__STREAM_DECODER_ABORTED
Jun 14 08:03 : player: played "http://192.168.178.31:49149/qobuz/track?version=1&trackId=614315"

On short tracks MPD works fine and no skips occur over hours or even days.

**Comment by MaxKellermann** [comment]
@nicolabotta it would be interesting to see what happens at the wire level. Can you make a wireshark dump, and analyze it?

**Comment by MaxKellermann** [comment]
Maybe `-f 'host 192.168.178.31'` to record only traffic to the HTTP server. Yes this will produce a lot of output, but the interesting part is only when the connection ends (and the time just before that).

**Comment by nicolabotta** [comment]
I have a dump and I can set filters and search for strings at times around 20:44

Jun 14 20:31 : player: played "http://192.168.178.101:49149/qobuz/track?version=1&trackId=47572575"
Jun 14 20:44 : exception: CURL failed: transfer closed with 14279665 bytes remaining to read
Jun 14 20:45 : player: played "http://192.168.178.101:49149/qobuz/track?version=1&trackId=47572576"

But what am I supposed to search for? How is a curl failure likely to show up in the analysis?

**Comment by nicolabotta** [comment]
![screenshot_2018-06-14_23-24-22](https://user-images.githubusercontent.com/2726946/41439404-ea53e0b6-702a-11e8-81e3-e81b8ed83fdc.png)

The blue line should be the request for the next track: it is the only occurrence of 47572577 in the dump. Somewhere above the error occurs. Perhaps in the red line?

**Comment by MaxKellermann** [comment]
You should follow the connection which receives the RST from the server (the red line).

**Comment by nicolabotta** [comment]
Following the TCP stream (at 20:44:56.219303561) which receives the RST immediately before the request for the next track (at 20:44:56.481889972) yields
![screenshot_2018-06-15_10-15-01](https://user-images.githubusercontent.com/2726946/41457864-4e7dbf80-7085-11e8-9d4a-f1861a4a4241.png)

**Comment by MaxKellermann** [comment]
Hm, that's not the right connection - but still I wonder why it receives a RST. The server declares it can do keep-alive, and then aborts the connection. The interesting part is why the data connection is being aborted by the server.

**Comment by nicolabotta** [comment]
There is another TCP connection short before that one (at time 20:44:56.178725583) that received a RST:
![screenshot_2018-06-15_10-58-09](https://user-images.githubusercontent.com/2726946/41459843-0f857b8c-708b-11e8-95a2-e1c6d5f36d27.png)
Following this other connection results in a huge list of packets, the last screenshot is
![screenshot_2018-06-15_11-58-05](https://user-images.githubusercontent.com/2726946/41462549-86a7cf1e-7093-11e8-83b3-11a805963eec.png)

**Comment by kingosticks** [comment]
Given the request timestamp in the first image is '1529001050' (14 June 2018 18:30:50), can you assume that 'etsp=1529004650' in the response url is the expiry timestamp (14 June 2018 19:30:50)?

**Comment by nicolabotta** [comment]
@kingosticks That would be a bit strange as the default keepalive time with curl is 60 seconds, not 60 minutes (https://ec.haxx.se/usingcurl-timeouts.html). But it is of course possible. I have tried different values of keepalive time (by setting them in .curlrc) with no obvious effects on the failures. 

What I have observed is that MPD 0.20.19 (used in the examples above, under Debian) behaves a little bit better than 0.19.21 under Raspbian. But it is not a day and night difference and streaming of long tracks often fails also in 0.20.19 as demonstrated above.
